File: benchmarks/String/string.pl
#!/usr/bin/perl

use strict;
use warnings;
use utf8;
use bytes;
use Time::HiRes qw( gettimeofday );

my $TARGET_SIZE = 1024 * 1024;
my $WORD_COUNT = 16;
my $LINE_LIMIT = 80;
my $FNV_OFFSET_HI = 0xcbf29ce4;
my $FNV_OFFSET_LO = 0x84222325;
my $FNV_PRIME_SMALL = 0x1b3;
my $REFERENCE_HASH_HI = 0xc6be9b92;
my $REFERENCE_HASH_LO = 0x67a2fb8e;

sub random_next {
	my ($state_ref) = @_;
	$$state_ref = (($$state_ref * 1664525) + 1013904223) & 0xffffffff;
	return $$state_ref;
}

sub multiply_hash {
	my ($hi, $lo) = @_;

	my $low_product = $lo * $FNV_PRIME_SMALL;
	my $new_lo = $low_product % 4294967296;
	my $carry = int($low_product / 4294967296);
	my $high_product =
		$hi * $FNV_PRIME_SMALL + $carry + (($lo << 8) & 0xffffffff);
	my $new_hi = $high_product % 4294967296;

	return ($new_hi, $new_lo);
}

sub hash_bytes {
	my ($text) = @_;
	my $hash_hi = $FNV_OFFSET_HI;
	my $hash_lo = $FNV_OFFSET_LO;
	my $i;

	for ($i = 0; $i < length($text); $i++) {
		$hash_lo = ($hash_lo ^ ord(substr($text, $i, 1))) & 0xffffffff;
		($hash_hi, $hash_lo) = multiply_hash($hash_hi, $hash_lo);
	}

	return ($hash_hi, $hash_lo);
}

sub build_source_text {
	my ($words_ref) = @_;
	my @parts = ();
	my $length = 0;
	my $word_count = 0;
	my $state = 0x12345678;

	while (1) {
		my $word = $words_ref->[random_next(\$state) % $WORD_COUNT];

		if ($word_count > 0) {
			push @parts, " ";
			$length += 1;
		}

		push @parts, $word;
		$length += length($word);
		$word_count++;

		if ($length > $TARGET_SIZE) {
			last;
		}
	}

	my $text = join("", @parts);
	my ($hash_hi, $hash_lo) = hash_bytes($text);
	return ($text, $word_count, $hash_hi, $hash_lo);
}

sub wrap_words {
	my ($words_ref) = @_;
	my @parts = ();
	my $line_length = 0;
	my $i;

	for ($i = 0; $i < scalar(@$words_ref); $i++) {
		my $word = $words_ref->[$i];
		my $word_length = length($word);

		if ($i == 0) {
			push @parts, $word;
			$line_length = $word_length;
			next;
		}

		if ($line_length + 1 + $word_length >= $LINE_LIMIT) {
			push @parts, "\n", $word;
			$line_length = $word_length;
		}
		else {
			push @parts, " ", $word;
			$line_length += 1 + $word_length;
		}
	}

	return join("", @parts);
}

sub hash_wrapped_text {
	my ($text) = @_;
	my @lines = split(/\n/, $text, -1);
	my $hash_hi = $FNV_OFFSET_HI;
	my $hash_lo = $FNV_OFFSET_LO;
	my ($i, $j);

	for ($i = 0; $i < scalar(@lines); $i++) {
		if ($i > 0) {
			$hash_lo = ($hash_lo ^ 32) & 0xffffffff;
			($hash_hi, $hash_lo) = multiply_hash($hash_hi, $hash_lo);
		}

		for ($j = 0; $j < length($lines[$i]); $j++) {
			$hash_lo =
				($hash_lo ^ ord(substr($lines[$i], $j, 1))) & 0xffffffff;
			($hash_hi, $hash_lo) = multiply_hash($hash_hi, $hash_lo);
		}
	}

	return ($hash_hi, $hash_lo);
}

sub format_hash {
	my ($hi, $lo) = @_;
	return sprintf("0x%08x%08x", $hi, $lo);
}

sub run_benchmark {
	my ($source_text, $source_hash_hi, $source_hash_lo, $runs) = @_;
	my $i;

	for ($i = 0; $i < $runs; $i++) {
		my @split_words = split(/ /, $source_text);
		my $wrapped_text = wrap_words(\@split_words);
		my ($wrapped_hi, $wrapped_lo) = hash_wrapped_text($wrapped_text);

		if (
			$wrapped_hi != $source_hash_hi ||
			$wrapped_lo != $source_hash_lo
		) {
			print STDERR
				"Hash mismatch: "
				. format_hash($wrapped_hi, $wrapped_lo)
				. " != "
				. format_hash($source_hash_hi, $source_hash_lo)
				. "\n";
			exit(1);
		}
	}
}

my @words = (
	"I",
	"we",
	"cat",
	"tree",
	"café",
	"naïve",
	"jalapeño",
	"mountain",
	"blueberry",
	"basketball",
	"grandfather",
	"microbiology",
	"determination",
	"responsibility",
	"experimentation",
	"counterclockwise"
);
my $runs = $#ARGV == -1 ? 1 : $ARGV[0];
my ($source_text, $source_word_count, $source_hash_hi, $source_hash_lo) =
	build_source_text(\@words);

if (
	$source_hash_hi != $REFERENCE_HASH_HI ||
	$source_hash_lo != $REFERENCE_HASH_LO
) {
	print STDERR
		"Source hash mismatch: "
		. format_hash($source_hash_hi, $source_hash_lo)
		. " != "
		. format_hash($REFERENCE_HASH_HI, $REFERENCE_HASH_LO)
		. "\n";
	exit(1);
}

my $begin = gettimeofday();
run_benchmark($source_text, $source_hash_hi, $source_hash_lo, $runs);
my $end = gettimeofday() - $begin;
printf STDERR "Perl: %0.3f s\n", $end;

File: benchmarks/String/string.lua
local TARGET_SIZE = 1024 * 1024
local WORD_COUNT = 16
local LINE_LIMIT = 80
local FNV_OFFSET_HI = 0xcbf29ce4
local FNV_OFFSET_LO = 0x84222325
local FNV_PRIME_SMALL = 0x1b3
local REFERENCE_HASH_HI = 0xc6be9b92
local REFERENCE_HASH_LO = 0x67a2fb8e

function random_next(state)
	state.value = (state.value * 1664525 + 1013904223) % 4294967296
	return state.value
end


function multiply_hash(hi, lo)
	local low_product = lo * FNV_PRIME_SMALL
	local new_lo = low_product % 4294967296
	local carry = math.floor(low_product / 4294967296)
	local high_product =
		hi * FNV_PRIME_SMALL + carry + ((lo * 256) % 4294967296)
	local new_hi = high_product % 4294967296

	return new_hi, new_lo
end


function hash_bytes(text)
	local hash_hi = FNV_OFFSET_HI
	local hash_lo = FNV_OFFSET_LO

	for i = 1, #text do
		hash_lo = (hash_lo ~ string.byte(text, i)) % 4294967296
		hash_hi, hash_lo = multiply_hash(hash_hi, hash_lo)
	end

	return hash_hi, hash_lo
end


function build_source_text(words)
	local parts = {}
	local length = 0
	local word_count = 0
	local state = { value = 0x12345678 }

	while true do
		local word = words[(random_next(state) % WORD_COUNT) + 1]

		if word_count > 0 then
			table.insert(parts, " ")
			length = length + 1
		end

		table.insert(parts, word)
		length = length + #word
		word_count = word_count + 1

		if length > TARGET_SIZE then
			break
		end
	end

	local text = table.concat(parts)
	local hash_hi, hash_lo = hash_bytes(text)
	return text, word_count, hash_hi, hash_lo
end


function wrap_words(words)
	local parts = {}
	local line_length = 0

	for i = 1, #words do
		local word = words[i]
		local word_length = #word

		if i == 1 then
			table.insert(parts, word)
			line_length = word_length
		elseif line_length + 1 + word_length >= LINE_LIMIT then
			table.insert(parts, "\n")
			table.insert(parts, word)
			line_length = word_length
		else
			table.insert(parts, " ")
			table.insert(parts, word)
			line_length = line_length + 1 + word_length
		end
	end

	return table.concat(parts)
end


function hash_wrapped_text(text)
	local hash_hi = FNV_OFFSET_HI
	local hash_lo = FNV_OFFSET_LO
	local first = true

	for line in string.gmatch(text, "[^\n]+") do
		if not first then
			hash_lo = (hash_lo ~ 32) % 4294967296
			hash_hi, hash_lo = multiply_hash(hash_hi, hash_lo)
		end
		first = false

		for i = 1, #line do
			hash_lo = (hash_lo ~ string.byte(line, i)) % 4294967296
			hash_hi, hash_lo = multiply_hash(hash_hi, hash_lo)
		end
	end

	return hash_hi, hash_lo
end


function format_hash(hi, lo)
	return string.format("0x%08x%08x", hi, lo)
end


function run_benchmark(source_text, source_hash_hi, source_hash_lo, runs)
	for i = 1, runs do
		local split_words = {}
		for word in string.gmatch(source_text, "[^ ]+") do
			table.insert(split_words, word)
		end
		local wrapped_text = wrap_words(split_words)
		local wrapped_hi, wrapped_lo = hash_wrapped_text(wrapped_text)

		if wrapped_hi ~= source_hash_hi or wrapped_lo ~= source_hash_lo then
			io.stderr:write(
				"Hash mismatch: "
				.. format_hash(wrapped_hi, wrapped_lo)
				.. " != "
				.. format_hash(source_hash_hi, source_hash_lo)
				.. "\n"
			)
			os.exit(1)
		end
	end
end

local words = {
	"I",
	"we",
	"cat",
	"tree",
	"café",
	"naïve",
	"jalapeño",
	"mountain",
	"blueberry",
	"basketball",
	"grandfather",
	"microbiology",
	"determination",
	"responsibility",
	"experimentation",
	"counterclockwise"
}
local runs = 1
if arg[1] ~= nil then
	runs = tonumber(arg[1]) or 1
end

local source_text, source_word_count, source_hash_hi, source_hash_lo =
	build_source_text(words)
if (
	source_hash_hi ~= REFERENCE_HASH_HI or
	source_hash_lo ~= REFERENCE_HASH_LO
) then
	io.stderr:write(
		"Source hash mismatch: "
		.. format_hash(source_hash_hi, source_hash_lo)
		.. " != "
		.. format_hash(REFERENCE_HASH_HI, REFERENCE_HASH_LO)
		.. "\n"
	)
	os.exit(1)
end

local t = os.clock()
run_benchmark(source_text, source_hash_hi, source_hash_lo, runs)
local diff = os.clock() - t
io.stderr:write(string.format("Lua: %.3f s\n", diff))
